opentelemetry-api/src/opentelemetry/context/propagation/tracecontexthttptextformat.py

```diff
@@ -22,18 +22,128 @@
 
 
 class TraceContextHTTPTextFormat(httptextformat.HTTPTextFormat):
-    """TODO: extracts and injects using w3c TraceContext's headers.
+    """Extracts and injects using w3c TraceContext's headers.
     """
 
+    _TRACEPARENT_HEADER_NAME = "traceparent"
+    _TRACESTATE_HEADER_NAME = "tracestate"
+    _TRACEPARENT_HEADER_FORMAT = (
+        "^[ \t]*([0-9a-f]{2})-([0-9a-f]{32})-([0-9a-f]{16})-([0-9a-f]{2})"
+        + "(-.*)?[ \t]*$"
+    )
+    _TRACEPARENT_HEADER_FORMAT_RE = re.compile(_TRACEPARENT_HEADER_FORMAT)
+
+    @classmethod
     def extract(
-        self, _get_from_carrier: httptextformat.Getter[_T], _carrier: _T
+        cls, get_from_carrier: httptextformat.Getter[_T], carrier: _T
     ) -> trace.SpanContext:
-        return trace.INVALID_SPAN_CONTEXT
+        """Extracts a valid SpanContext from the carrier.
+
+        If a header
+        """
+        header = get_from_carrier(carrier, cls._TRACEPARENT_HEADER_NAME)
+
+        if not header:
+            return trace.INVALID_SPAN_CONTEXT
+
+        match = re.search(cls._TRACEPARENT_HEADER_FORMAT_RE, header[0])
+        if not match:
+            return trace.INVALID_SPAN_CONTEXT
+
+        version = match.group(1)
+        trace_id = match.group(2)
+        span_id = match.group(3)
+        trace_options = match.group(4)
 
+        if trace_id == "0" * 32 or span_id == "0" * 16:
```

Comment:
I don't think this check is very useful here. I'd only check after constructing `span_context` at the end, that way we can do integer comparisons.

Reply:
That seems ok. The one point there is we have to reject all values if either is invalid, as well as this skips some level of processing. But should be fine.